src/soter/openssl/soter_ecdsa_common.c

```diff
@@ -22,28 +22,59 @@
 #include "soter/openssl/soter_engine.h"
 #include "soter/soter_ec_key.h"
 
-soter_status_t soter_ec_gen_key(EVP_PKEY_CTX* pkey_ctx)
+soter_status_t soter_ec_gen_key(EVP_PKEY** ppkey)
 {
-    EVP_PKEY* pkey;
-    EC_KEY* ec;
-    if (!pkey_ctx) {
+    soter_status_t res = SOTER_FAIL;
+    EVP_PKEY* param = NULL;
+    EVP_PKEY_CTX* param_ctx = NULL;
+    EVP_PKEY_CTX* pkey_ctx = NULL;
+
+    if (!ppkey) {
         return SOTER_INVALID_PARAMETER;
     }
-    pkey = EVP_PKEY_CTX_get0_pkey(pkey_ctx);
-    if (!pkey) {
-        return SOTER_INVALID_PARAMETER;
+
+    param_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+    if (!param_ctx) {
+        res = SOTER_NO_MEMORY;
+        goto err;
     }
-    if (EVP_PKEY_EC != EVP_PKEY_id(pkey)) {
-        return SOTER_INVALID_PARAMETER;
+
+    if (EVP_PKEY_paramgen_init(param_ctx) != 1) {
+        res = SOTER_FAIL;
+        goto err;
     }
-    ec = EVP_PKEY_get0(pkey);
-    if (NULL == ec) {
-        return SOTER_INVALID_PARAMETER;
+    if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(param_ctx, NID_X9_62_prime256v1) != 1) {
+        res = SOTER_FAIL;
+        goto err;
     }
-    if (1 == EC_KEY_generate_key(ec)) {
-        return SOTER_SUCCESS;
+    if (EVP_PKEY_paramgen(param_ctx, &param) != 1) {
+        res = SOTER_FAIL;
+        goto err;
     }
-    return SOTER_FAIL;
+
+    pkey_ctx = EVP_PKEY_CTX_new(param, NULL);
+    if (!pkey_ctx) {
+        res = SOTER_NO_MEMORY;
+        goto err;
+    }
+
+    if (EVP_PKEY_keygen_init(pkey_ctx) != 1) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+    if (EVP_PKEY_keygen(pkey_ctx, ppkey) != 1) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    res = SOTER_SUCCESS;
+
+err:
+    EVP_PKEY_CTX_free(param_ctx);
```

Comment:
I assume these functions check for `NULL` internally, right?

Reply:
> I assume these functions check for NULL internally, right?

That is so. They follow the free() convention: NULL argument is a no-op.